### DISCUSSION TITLE
Is it possible to put the NativeMenu of a window into a separate  control？

### DISCUSSION THREAD

**Discussion by beginor** [AvaloniaUI/Avalonia#15560] [Q&A] [answered]
Since we can define NativeMenu for a window like this [Native Window Sample](https://github.com/AvaloniaUI/Avalonia/blob/ceb172f22790b74c56550889b02eb7e8f1dbf8d1/samples/IntegrationTestApp/MainWindow.axaml#L24)

If the window has too many menu items,  there will be lots of code to handle menu click in the window.cs file, and I think this is not good for maintaince. 

So my question is:  is it possible to put the native menu defs of a window into a seperate control file?

**Comment by maxkatz6** [comment, marked as answer]
NativeMenu itself can only be set on Window.

But as with many other controls, you can extract it into a separated XAML file anyway:

XAML file MyMenu.axaml:
```xaml
<NativeMenu xmlns="https://github.com/avaloniaui"
            xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
            x:Class="Sandbox.MyMenu">
  <NativeMenuItem Header="File" />
</NativeMenu>
```

And code behind for this file:
```csharp
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace Sandbox;

public partial class MyMenu : NativeMenu
{
    public MyMenu()
    {
        AvaloniaXamlLoader.Load(this);
    }
}
```

Which then can be used like this:
```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        xmlns:sandbox="using:Sandbox"
        x:Class="Sandbox.MainWindow">
  <NativeMenu.Menu>
    <sandbox:MyMenu />
  </NativeMenu.Menu>

  <DockPanel>
    <NativeMenuBar DockPanel.Dock="Top" />
  </DockPanel>
</Window>
```

**Reply by maxkatz6** [reply]
In general, it follows the same rules as with resource dictionaries and other XAML elements that can be extracted to a separated class with x:Class defined on them.

**Reply by beginor** [reply]
OK, it works, thanks very much! I have not use  `AvaloniaXamlLoader.Load(this)` before.

**Reply by maxkatz6** [reply]
Normally, this method is called from the inside of InitializeComponent() method.

**Reply by beginor** [reply]
But `NativeMenu` does not have DataContext property, so I can't use databindings like other control?

**Reply by maxkatz6** [reply]
Yes, I don't think you can use bindings there.

**Reply by beginor** [reply]
Ok, then I will try the `Menu` and `MenuItem` class instead. anyway, thanks for your kind replies.
